lib/mcp/server/transports/streamable_http_transport.rb

```diff
@@ -0,0 +1,295 @@
+                rescue IOError, Errno::EPIPE => e
+                  MCP.configuration.exception_reporter.call(
+                    e,
+                    { session_id: sid, error: "Failed to send notification" },
+                  )
+                  failed_sessions << sid
+                end
+              end
+
+              # Clean up failed sessions
+              failed_sessions.each { |sid| cleanup_session_unsafe(sid) }
+
+              sent_count
+            end
+          end
+        end
+
+        private
+
+        def send_to_stream(stream, data)
+          message = data.is_a?(String) ? data : data.to_json
+          stream.write("data: #{message}\n\n")
+          stream.flush if stream.respond_to?(:flush)
+        end
+
+        def send_ping_to_stream(stream)
+          stream.write(": ping #{Time.now.iso8601}\n\n")
+          stream.flush if stream.respond_to?(:flush)
+        end
+
+        def handle_post(request)
+          body_string = request.body.read
+          session_id = extract_session_id(request)
+
+          body = parse_request_body(body_string)
+          return body unless body.is_a?(Hash) # Error response
+
+          if body["method"] == "initialize"
+            handle_initialization(body_string, body)
+          else
+            handle_regular_request(body_string, session_id)
+          end
+        rescue StandardError => e
+          ModelContextProtocol.configuration.exception_reporter.call(e, { request: body_string })
+          [500, { "Content-Type" => "application/json" }, [{ error: "Internal server error" }.to_json]]
+        end
+
+        def handle_get(request)
+          session_id = extract_session_id(request)
+
+          return missing_session_id_response unless session_id
+          return session_not_found_response unless session_exists?(session_id)
+
+          setup_sse_stream(session_id)
+        end
+
+        def handle_delete(request)
+          session_id = request.env["HTTP_MCP_SESSION_ID"]
+
+          return [
+            400,
+            { "Content-Type" => "application/json" },
+            [{ error: "Missing session ID" }.to_json],
+          ] unless session_id
+
+          cleanup_session(session_id)
+          [200, { "Content-Type" => "application/json" }, [{ success: true }.to_json]]
+        end
+
+        def cleanup_session(session_id)
+          @mutex.synchronize do
+            cleanup_session_unsafe(session_id)
+          end
+        end
+
+        def cleanup_session_unsafe(session_id)
+          session = @sessions[session_id]
+          return unless session
+
+          begin
+            session[:stream]&.close
+          rescue
+            nil
+          end
+          @sessions.delete(session_id)
+        end
+
+        def extract_session_id(request)
+          request.env["HTTP_MCP_SESSION_ID"]
+        end
+
+        def parse_request_body(body_string)
+          JSON.parse(body_string)
+        rescue JSON::ParserError
```

Comment:
```suggestion
        rescue JSON::ParserError, TypeError
```
rescue on `TypeError` would address issues with nil request bodies.